### PR TITLE
ISSUE-1023 Provide HTTP API endpoints for merging component catalog information and metrics

### PR DESCRIPTION
**Pull request by HeartSaVioR**
* add API endpoints (common parameters: "from", "to")
  * /api/v1/catalog/topologies/:topologyId/metrics
  * /api/v1/catalog/topologies/:topologyId/sources/metrics
  *  /api/v1/catalog/topologies/:topologyId/sources/:sourceId/metrics
  *  /api/v1/catalog/topologies/:topologyId/processors/metrics
  *  /api/v1/catalog/topologies/:topologyId/processors/:processorId/metrics
  *  /api/v1/catalog/topologies/:topologyId/sinks/metrics
  *  /api/v1/catalog/topologies/:topologyId/sinks/:sinkId/metrics
* above API will provide the following response
  * catalog information for topology or component
  * summarized metrics (current time window, previous time window)
  * time-series metrics

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
This is preliminary versions of PR for #1023 to check we are on same page on backend works.

Please note that `error count` is not reflected: we talked about trying out querying to Solr with WARN/ERROR/FATAL. For the change #1002 is needed. Let's take up later after #1002 is merged in.
If we are OK with showing `failed count` as `error count`, that's available in providing APIs.

The API endpoints are multi-purpose: covers both `Overview` and `Metric`. Please let me know if we want to split API into two to reduce the response.

**Comment by HeartSaVioR** [comment]
Example of responses:

> /api/v1/viewmode/topologies/2?from=1511337448793&to=1511337898793

```
{
   "topology":{
      "id":2,
      "versionId":6,
      "name":"SimpleTopology",
      "description":null,
      "namespaceId":2,
      "config":"{\"topology.workers\":1,\"topology.acker.executors\":1,\"topology.message.timeout.secs\":30,\"clustersSecurityConfig\":[{}]}",
      "timestamp":1511336920661
   },
   "overviewMetrics":{
      "emitted":19081200,
      "acked":19122460,
      "failed":0,
      "latency":281.60972280307584,
      "prevEmitted":39829300,
      "prevAcked":39779220,
      "prevFailed":0,
      "prevLatency":44.51442292550002
   },
   "timeSeriesMetrics":{
      "componentName":"SimpleTopology",
      "inputRecords":{
         "1511337450000":2705740.0,
         "1511337480000":2701840.0,
                             ...
         "1511337840000":325790.0,
         "1511337870000":328020.0
      },
      "outputRecords":{
         "1511337450000":3247400.0,
         "1511337480000":3242110.0,
                             ...
         "1511337840000":390910.0,
         "1511337870000":393580.0
      },
      "failedRecords":{

      },
      "processedTime":{
         "1511337450000":0.11608099331927946,
         "1511337480000":0.11963310603824166,
                                 ...
         "1511337840000":0.2533354405040704,
         "1511337870000":0.24892076565877344
      },
      "recordsInWaitQueue":{
         "1511337450000":10.833333333333334,
         "1511337480000":11.083333333333334,
                                 ...
         "1511337840000":98.5,
         "1511337870000":101.5
      },
      "misc":{
         "ackedRecords":{
            "1511337450000":3246740.0,
            "1511337480000":3242190.0,
                                 ...
            "1511337840000":390920.0,
            "1511337870000":393580.0
         },
         "completeLatency":{
            "1511337450000":47.286902517467006,
            "1511337480000":47.15573495088157,
                                  ...
            "1511337840000":414.77755530097323,
            "1511337870000":412.0619091186337
         }
      }
   }
}
```

> /api/v1/viewmode/topologies/2/sources?from=1511337448793&to=1511337898793

```
{
   "entities":[
      {
         "component":{
            "id":1,
            "topologyId":2,
            "topologyComponentBundleId":1,
            "versionId":6,
            "name":"KAFKA",
            "description":"",
            "config":{
               "properties":{
                  "cluster":"2",
                  "consumerGroupId":"truck_speed_events_avro",
                  "retryDelayPeriodMs":2,
                  "maximumUncommittedOffsets":10000000,
                  "maxRecordsPerPoll":"500",
                  "maximumRetries":2147483647,
                  "parallelism":1,
                  "emitNullTuples":false,
                  "partitionRefreshPeriodMs":2000,
                  "securityProtocol":"PLAINTEXT",
                  "bootstrapServers":"..............",
                  "retryDelayMaximumMs":10000,
                  "fetchMaximumBytesPerPartition":"1048576",
                  "firstPollOffsetStrategy":"UNCOMMITTED_EARLIEST",
                  "sslTruststoreType":"JKS",
                  "clusters":"cl1",
                  "fetchMinimumBytes":"1",
                  "sslKeystoreType":"JKS",
                  "sslTrustManagerAlgo":"PKIX",
                  "retryInitialDelayMs":0,
                  "sslKeyManagerAlgo":"SunX509",
                  "sslProtocol":"TLS",
                  "topic":"truck_speed_events_avro",
                  "readerSchemaVersion":"1",
                  "pollTimeoutMs":200,
                  "offsetCommitPeriodMs":30000
               }
            },
            "reconfigure":false,
            "outputStreams":[
               {
                  "id":4,
                  "versionId":6,
                  "streamId":"kafka_stream_9",
                  "description":null,
                  "topologyId":2,
                  "fields":[
                     {
                        "name":"driverId",
                        "type":"LONG",
                        "optional":false
                     },
                     {
                        "name":"truckId",
                        "type":"LONG",
                        "optional":false
                     },
                     {
                        "name":"eventTime",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"eventType",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"latitude",
                        "type":"DOUBLE",
                        "optional":false
                     },
                     {
                        "name":"longitude",
                        "type":"DOUBLE",
                        "optional":false
                     },
                     {
                        "name":"eventKey",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"correlationId",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"driverName",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"routeId",
                        "type":"LONG",
                        "optional":false
                     },
                     {
                        "name":"routeName",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"eventDate",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"miles",
                        "type":"LONG",
                        "optional":false
                     }
                  ],
                  "timestamp":1511336920661
               }
            ]
         },
         "overviewMetrics":{
            "emitted":6389010,
            "acked":3194510,
            "failed":0,
            "latency":281.60972280307584,
            "prevEmitted":13243390,
            "prevAcked":6620500,
            "prevFailed":0,
            "prevLatency":44.51442292550002
         },
         "timeSeriesMetrics":{
            "componentName":"KAFKA",
            "inputRecords":{

            },
            "outputRecords":{
               "1511337450000":1082060.0,
               "1511337480000":1080770.0,
...
               "1511337840000":130310.0,
               "1511337870000":131140.0
            },
            "failedRecords":{

            },
            "processedTime":{

            },
            "recordsInWaitQueue":{
               "1511337450000":2.0,
               "1511337480000":2.0,
...
               "1511337840000":3.0,
               "1511337870000":3.0
            },
            "misc":{
               "ackedRecords":{
                  "1511337450000":541020.0,
                  "1511337480000":540380.0,
...
                  "1511337840000":65150.0,
                  "1511337870000":65580.0
               },
               "completeLatency":{
                  "1511337450000":47.286902517467006,
                  "1511337480000":47.15573495088157,
...
                  "1511337840000":414.77755530097323,
                  "1511337870000":412.0619091186337
               }
            }
         }
      }
   ]
}
```

> /api/v1/viewmode/topologies/2/processors?from=1511337448793&to=1511337898793

```
{
   "entities":[
      {
         "component":{
            "id":2,
            "topologyId":2,
            "topologyComponentBundleId":4,
            "versionId":6,
            "name":"RULE",
            "description":"",
            "config":{
               "properties":{
                  "rules":[
                     2
                  ]
               }
            },
            "reconfigure":false,
            "outputStreams":[
               {
                  "id":5,
                  "versionId":6,
                  "streamId":"rule_transform_stream_1",
                  "description":null,
                  "topologyId":2,
                  "fields":[
                     {
                        "name":"driverId",
                        "type":"LONG",
                        "optional":false
                     },
                     {
                        "name":"truckId",
                        "type":"LONG",
                        "optional":false
                     },
                     {
                        "name":"eventTime",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"eventType",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"latitude",
                        "type":"DOUBLE",
                        "optional":false
                     },
                     {
                        "name":"longitude",
                        "type":"DOUBLE",
                        "optional":false
                     },
                     {
                        "name":"eventKey",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"correlationId",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"driverName",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"routeId",
                        "type":"LONG",
                        "optional":false
                     },
                     {
                        "name":"routeName",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"eventDate",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"miles",
                        "type":"LONG",
                        "optional":false
                     }
                  ],
                  "timestamp":1511336920661
               },
               {
                  "id":6,
                  "versionId":6,
                  "streamId":"rule_notifier_stream_1",
                  "description":null,
                  "topologyId":2,
                  "fields":[
                     {
                        "name":"driverId",
                        "type":"LONG",
                        "optional":false
                     },
                     {
                        "name":"truckId",
                        "type":"LONG",
                        "optional":false
                     },
                     {
                        "name":"eventTime",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"eventType",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"latitude",
                        "type":"DOUBLE",
                        "optional":false
                     },
                     {
                        "name":"longitude",
                        "type":"DOUBLE",
                        "optional":false
                     },
                     {
                        "name":"eventKey",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"correlationId",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"driverName",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"routeId",
                        "type":"LONG",
                        "optional":false
                     },
                     {
                        "name":"routeName",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"eventDate",
                        "type":"STRING",
                        "optional":false
                     },
                     {
                        "name":"miles",
                        "type":"LONG",
                        "optional":false
                     }
                  ],
                  "timestamp":1511336920661
               }
            ]
         },
         "overviewMetrics":{
            "emitted":6304260,
            "acked":3152120,
            "failed":0,
            "latency":0.5958750001119716,
            "prevEmitted":13342270,
            "prevAcked":6671270,
            "prevFailed":0,
            "prevLatency":0.015949717576625392
         },
         "timeSeriesMetrics":{
            "componentName":"RULE",
            "inputRecords":{
               "1511337450000":541640.0,
               "1511337480000":540290.0,
...
               "1511337840000":65150.0,
               "1511337870000":65660.0
            },
            "outputRecords":{
               "1511337450000":1083280.0,
               "1511337480000":1080550.0,
...
               "1511337840000":130290.0,
               "1511337870000":131300.0
            },
            "failedRecords":{

            },
            "processedTime":{
               "1511337450000":0.015989069827554374,
               "1511337480000":0.015657725621462643,
...
               "1511337840000":0.9179116258221923,
               "1511337870000":0.9141029546146817
            },
            "recordsInWaitQueue":{
               "1511337450000":2.0,
               "1511337480000":2.0,
...
               "1511337840000":580.0,
               "1511337870000":598.0
            },
            "misc":{
               "ackedRecords":{
                  "1511337450000":541620.0,
                  "1511337480000":540280.0,
...
                  "1511337840000":65150.0,
                  "1511337870000":65660.0
               }
            }
         }
      }
   ]
}
```

> /api/v1/viewmode/topologies/2/sinks?from=1511337448793&to=1511337898793

```
{
   "entities":[
      {
         "component":{
            "id":3,
            "topologyId":2,
            "topologyComponentBundleId":19,
            "versionId":6,
            "name":"KAFKA-1",
            "description":"",
            "config":{
               "properties":{
                  "cluster":"2",
                  "parallelism":1,
                  "compressionType":"none",
                  "keySerializer":"ByteArray",
                  "timeout":"30000",
                  "securityProtocol":"PLAINTEXT",
                  "bootstrapServers":"...",
                  "metadataMaxAge":"300000",
                  "sslTruststoreType":"JKS",
                  "requestTimeout":"30000",
                  "clusters":"cl1",
                  "metadataFetchTimeout":"60000",
                  "clientId":"",
                  "maxRequestSize":"1048576",
                  "reconnectBackoff":"50",
                  "blocKOnBufferFull":true,
                  "sendBufferSize":"131072",
                  "sslKeystoreType":"JKS",
                  "lingerTime":"0",
                  "sslTrustManagerAlgo":"PKIX",
                  "retryBackoff":"100",
                  "sslKeyManagerAlgo":"SunX509",
                  "serProtocolVersion":1,
                  "maxInflighRequests":"5",
                  "async":true,
                  "retries":"0",
                  "bufferMemory":"33554432",
                  "writerSchemaVersion":"1",
                  "receiveBufferSize":"32768",
                  "fireAndForget":false,
                  "ackMode":"Leader",
                  "sslProtocol":"TLS",
                  "topic":"truck_speed_events_avro",
                  "batchSize":"16384",
                  "maxConnectionIdle":"540000",
                  "maxBlock":"60000"
               }
            },
            "reconfigure":false
         },
         "overviewMetrics":{
            "emitted":3193370,
            "acked":3193290,
            "failed":0,
            "latency":0.6309365674035676,
            "prevEmitted":6622580,
            "prevAcked":6622590,
            "prevFailed":0,
            "prevLatency":0.661444385878118
         },
         "timeSeriesMetrics":{
            "componentName":"KAFKA-1",
            "inputRecords":{
               "1511337450000":541040.0,
               "1511337480000":540400.0,
...
               "1511337840000":65150.0,
               "1511337870000":65580.0
            },
            "outputRecords":{
               "1511337450000":541040.0,
               "1511337480000":540400.0,
...
               "1511337840000":65160.0,
               "1511337870000":65580.0
            },
            "failedRecords":{

            },
            "processedTime":{
               "1511337450000":0.6736165021625818,
               "1511337480000":0.6952956709501166,
...
               "1511337840000":0.5911798796425004,
               "1511337870000":0.569685879841415
            },
            "recordsInWaitQueue":{
               "1511337450000":55.0,
               "1511337480000":56.5,
...
               "1511337840000":2.0,
               "1511337870000":2.0
            },
            "misc":{
               "ackedRecords":{
                  "1511337450000":541020.0,
                  "1511337480000":540380.0,
...
                  "1511337840000":65140.0,
                  "1511337870000":65580.0
               }
            }
         }
      }
   ]
}
```

**Comment by HeartSaVioR** [comment]
Rebased.

**Comment by arunmahadevan** [comment]
@HeartSaVioR , instead of "/api/v1/viewmode/topologies/:topologyid" may be better to nest all these under respective components e.g. "/api/v1/catalog/topologies/:topologyid/detail".

**Comment by HeartSaVioR** [comment]
Rebased.

**Comment by arunmahadevan** [comment]
@HeartSaVioR , can you update https://github.com/hortonworks/streamline/pull/1032#issue-276014344 to reflect the new endpoints ?

**Comment by HeartSaVioR** [comment]
@arunmahadevan Updated.

File: streams/catalog/src/main/java/com/hortonworks/streamline/streams/catalog/CatalogToLayoutConverter.java [hunk 1]
```diff
@@ -17,12 +17,10 @@
 
 import com.hortonworks.streamline.streams.catalog.Topology;
 import com.hortonworks.streamline.streams.catalog.TopologyComponent;
-import com.hortonworks.streamline.streams.layout.component.StreamlineComponent;
-import com.hortonworks.streamline.streams.layout.component.TopologyDag;
-import com.hortonworks.streamline.streams.layout.component.TopologyDagVisitor;
-import com.hortonworks.streamline.streams.layout.component.TopologyLayout;
+import com.hortonworks.streamline.streams.layout.component.*;
```
Comment (arunmahadevan) [inline streams/catalog/src/main/java/com/hortonworks/streamline/streams/catalog/CatalogToLayoutConverter.java]: wildcard

File: streams/runners/storm/metrics/src/main/java/com/hortonworks/streamline/streams/metrics/storm/topology/StormTopologyTimeSeriesMetricsImpl.java [hunk 2]
```diff
@@ -18,18 +18,14 @@
...
 import com.hortonworks.streamline.streams.storm.common.StormRestAPIClient;
 import com.hortonworks.streamline.streams.storm.common.StormTopologyUtil;
 
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.TreeMap;
+import java.util.*;
```
Comment (arunmahadevan) [inline streams/runners/storm/metrics/src/main/java/com/hortonworks/streamline/streams/metrics/storm/topology/StormTopologyTimeSeriesMetricsImpl.java]: wildcard

File: streams/service/src/main/java/com/hortonworks/streamline/streams/service/TopologyViewModeResource.java [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+import com.hortonworks.streamline.streams.catalog.*;
+import com.hortonworks.streamline.streams.catalog.service.StreamCatalogService;
+import com.hortonworks.streamline.streams.metrics.topology.TopologyTimeSeriesMetrics;
+import com.hortonworks.streamline.streams.metrics.topology.service.TopologyMetricsService;
+import com.hortonworks.streamline.streams.security.Roles;
+import com.hortonworks.streamline.streams.security.SecurityUtil;
+import com.hortonworks.streamline.streams.security.StreamlineAuthorizer;
+import org.jooq.lambda.Unchecked;
+
+import javax.ws.rs.*;
```
Comment (arunmahadevan) [inline streams/service/src/main/java/com/hortonworks/streamline/streams/service/TopologyViewModeResource.java]: wildcard

File: streams/service/src/main/java/com/hortonworks/streamline/streams/service/TopologyViewModeResource.java [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+import javax.ws.rs.core.*;
```
Comment (arunmahadevan) [inline streams/service/src/main/java/com/hortonworks/streamline/streams/service/TopologyViewModeResource.java]: wildcard

**Comment by HeartSaVioR** [comment]
@arunmahadevan Addressed.

**Comment by arunmahadevan** [comment]
+1

**Pull request closed without merging**